pkg/controllers/ratelimitpolicy/ratelimitpolicy_controller.go

```diff
@@ -0,0 +1,257 @@
+		return nil
+	}
+
+	err = controllerutil.SetControllerReference(targetGateway, rateLimitPolicy, r.Scheme)
+	if err != nil {
+		return err
+	}
+
+	return err
+}
+
+// getAllClusters returns an array of all cluster secrets available.
+func (r *RateLimitPolicyReconciler) getAllClusters(ctx context.Context) ([]corev1.Secret, error) {
+	clusterList := &corev1.SecretList{}
+	listOptions := client.MatchingLabels{
+		"argocd.argoproj.io/secret-type": "cluster",
+	}
+	err := r.Client.List(ctx, clusterList, listOptions)
+	if err := client.IgnoreNotFound(err); err != nil {
+		log.Log.Error(err, "Unable to fetch cluster Secrets")
+		return clusterList.Items, err
+	}
+	return clusterList.Items, nil
+}
+
+// getClustersFromAnnotations returns an array of cluster secrets based on the given objects sync annotations.
+// If the wildcard `all` annotation is present, all cluster secrets available are returned.
+func (r *RateLimitPolicyReconciler) getClustersFromAnnotations(ctx context.Context, obj metav1.Object) ([]corev1.Secret, error) {
+	clusters := []corev1.Secret{}
+
+	allClusters, err := r.getAllClusters(ctx)
+	if err != nil {
+		return clusters, err
+	}
+
+	if metadata.HasAnnotation(obj, syncer.MCTC_SYNC_ANNOTATION_PREFIX+syncer.MCTC_SYNC_ANNOTATION_WILDCARD) {
+		return allClusters, nil
+	}
+
+	clustersMap := map[string]corev1.Secret{}
+	for _, cluster := range allClusters {
+		clustersMap[cluster.Name] = cluster
+	}
+	for annKey := range obj.GetAnnotations() {
+		if strings.HasPrefix(annKey, syncer.MCTC_SYNC_ANNOTATION_PREFIX) {
+			_, clusterName, found := strings.Cut(annKey, syncer.MCTC_SYNC_ANNOTATION_PREFIX)
+			if found {
+				if cluster, exists := clustersMap[clusterName]; exists {
+					clusters = append(clusters, cluster)
+				}
+			}
+		}
+	}
+	return clusters, nil
+}
+
+// setSyncAnnotationsForClusters adds sync annotations to the given resource for each cluster.
+// Note: A sync annotation is added for each individual cluster even if the wildcard `all` annotation was used on the target Gateway.
+// We need to add patch annotations for each cluster individually, so for consistency sync annotations are the same.
+func setSyncAnnotationsForClusters(obj metav1.Object, clusters []corev1.Secret) {
+	annotations := map[string]string{}
+	// keep all non sync related annotations
+	for annKey, annValue := range obj.GetAnnotations() {
+		if !strings.HasPrefix(annKey, syncer.MCTC_SYNC_ANNOTATION_PREFIX) {
+			annotations[annKey] = annValue
+		}
+	}
+	// add all sync annotations
+	for _, cluster := range clusters {
+		annotations[syncer.MCTC_SYNC_ANNOTATION_PREFIX+cluster.Name] = "true"
+	}
+
+	obj.SetAnnotations(annotations)
+}
+
+// setPatchAnnotationsForClusters adds patch annotations to the given RateLimitPolicy for each cluster.
+// Adds a patch annotation to the RLP that injects an Action with a generic key (key: cluster name, value: 1) which
+// can then be used in limit conditions.
+// ToDo add attributes from an annotation on the cluster secrets
```

Comment:
@maleck13 fyi, the reason I'm referring to the secrets everywhere  and not relying on just the name is so we can do this. It would be fairly straightforward to add, but did you have an idea of what you wanted this to look like? Was the idea to add a single annotation with a bob of json ( i.e. {"provider": "aws"}), or multiple annotations with a single key/value that we merge together?

Reply:
I had thought of it as multiple annotations. ```kuadrant.io/attribute-cloud: aws kuadrant.io/attribute-geo: EU-West``` so basically a convention